Make after & else clauses optional

nathanalderson opened the pull request:
This PR makes the `after` and `else` clauses optional. Essentially, this:

```elixir
retry with: ... do
  ...
end
```

Is now equivalent to:

```elixir
retry with: ... do
  ...
after
  result -> result
else
  error -> raise error
end
```

I could be convinced to change the default `else` behavior to simply returning the exception, but re-raising it seems more natural to me.

I also improved the error messaging slightly by calling out the specific option or clause the user messed up when possible. So you'll get a message like `you must provide a "do" clause` or `option "foo" is unsupported` instead of a more generic message.

So far I have only done the implementation for the `retry` macro, but if you like the concept, I will update `wait` similarly. I think `retry_while` is already fine in this respect since it only has a `do` block.

safwank commented:
Thanks for the PR @nathanalderson! I've been wanting to do this for a while. I'll take a look at the PR this weekend and get back to you.

nathanalderson commented:
Went ahead and did `wait`. This:

```elixir
wait ... do
  ...
end
```

Is now equivalent to this:

```elixir
wait ... do
  ...
after
  result -> {:ok, result}
else
  error -> {:error, error}
end
```

A few other notes:

* The order of the clauses no longer matters.
* We don't detect and error on duplicate clauses. The last one wins (like a standard keyword list). Could add a check for this if desired.

nathanalderson commented:
After some thought, I realized that the appropriate thing to do in the default `retry ... else` clause is to reraise if the last failure was an exception and return the failed value otherwise. That's now the behavior. In other words:

```elixir
retry with: ... do
  ...
end
```

Is now equivalent to:

```elixir
retry with: ... do
  ...
after
  result -> result
else
  e when is_exception(e) -> raise e
  e -> e
end
```

nathanalderson commented:
Sorry, meant to mark this as ready. Let me know if you have feedback.

I have a follow-up PR that addresses some more needs I have including some discussed in #54.

nathanalderson commented:
Just bumping this a bit. Not super urgent but didn't want it to fall off your radar or something. Let me know if there's anything I can do to assist.

safwank commented:
@nathanalderson Rest assured I haven't forgotten about this. I'm currently busy with life, so please bear with me.

nathanalderson commented:
Periodic bump. Would love to get this in and introduce a follow-up. 🙂

safwank commented on lib/retry.ex:
```diff
@@ -71,15 +127,14 @@ defmodule Retry do
       end
 
+  The `after` and `else` clauses are optional. By default, a successful value is just returned. If
+  the timeout is expires, the last erroneous value is returned or the last exception is re-raised.
```
Typo in `If the timeout ~is~ expires`

safwank replied:
Can you also include sample code without `after` and `else`?

safwank commented on test/retry_test.exs:
```diff
@@ -399,39 +449,43 @@ defmodule RetryTest do
 
-      assert_raise ArgumentError, error_message, fn ->
-        Code.eval_string(
-          "wait [1, 2, 3], do: :valid, after: :valid, after: :duplicate",
```
Any reason why you removed the tests for duplicate clauses?

nathanalderson replied:
I had removed handling for that case at one point and forgot to re-add it. I've corrected it now.

safwank commented:
@nathanalderson Thanks for the fixes. A couple of comments
- The test coverage has dropped a bit. Mind plugging the gap?
- Can you please revert [this](https://github.com/safwank/ElixirRetry/pull/55/commits/10ef209a34165ba180fe44b146b033f6cd4c95a7) commit and create a separate PR for it so we can keep this PR small and focused?

nathanalderson commented:
Sorry, the merge commit was accidental. That's the follow-up PR I've mentioned. Meant to merge `upstream/master` instead of `origin/master`. I've reverted. I also covered the two missed syntax errors in unit test.

I also added a comment to disable the complexity warning from credo on the main `retry` macro. I tried moving the else/after logic into a function and that did fix the warning, but in my opinion it made things less readable. I'm open to doing that if you prefer.

safwank reviewed (approved):
🚀
